File: backend/app/schemas/icp.py
"""ICP schemas + prospect-import history schemas.

Two unrelated things share this module for historical reasons (the name
predates the split, and ImportBatchOut has many callers):

- IcpBase/IcpCreate/IcpOut back the ICP CRUD API (routes/icp.py). These were
  deleted in 2ba62a9 when ICP-as-a-scoring-filter was removed, and are
  restored here for ICP-as-a-discovery-seed - a different feature with the
  same storage. IcpCompaniesOut is deliberately NOT restored: it backed
  "which companies match this ICP", which is exactly the filtering semantics
  the product moved away from.
- ImportBatchOut is prospect-upload history, which has no ICP dependency and
  is served by controllers/icp_imports.py.
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class ImportBatchOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    import_batch_id: UUID
    workspace_id: UUID | None = None
    # 'upload' (a person supplied a file) or 'generated' (discovered from an
    # ICP). Generated batches set icp_id to the ICP they came from; uploads
    # leave it null.
    source: str = "upload"
    icp_id: UUID | None = None
    icp_name: str | None = None
    file_names: list[str] | None = None
    files_processed: int
    total_rows: int
    companies_ingested: int
    signals_extracted: int
    # New pipeline sales-status counts (brief section 7).
    sales_ready_count: int = 0
    high_priority_count: int = 0
    warm_count: int = 0
    monitor_count: int = 0
    low_priority_count: int = 0
    # Legacy read-only counters (old ICP/gate pipeline).
    matched_icp_count: int = 0
    active_count: int = 0
    nurture_count: int = 0
    # 'pending' while background research+scoring is still running (see
    # excel_pipeline.score_companies_in_background) - counts are 0 until this
    # flips to 'complete'.
    scoring_status: str = "complete"
    # Operational status of the background task (brief items 21, 23):
    # pending | complete | complete_with_warnings | failed.
    research_status: str = "pending"
    companies_researched: int = 0
    research_failure_count: int = 0
    llm_failure_count: int = 0
    scoring_failure_count: int = 0
    processing_started_at: datetime | None = None
    processing_completed_at: datetime | None = None
    processing_error: str | None = None
    processing_warnings: list | dict | None = None
    created_at: datetime | None = None


class IcpBase(BaseModel):
    """The ICP criteria a user can set. Shared by create (POST) and
    full-replace update (PUT) - both accept exactly the same fields, so an
    omitted field means "no constraint on this criterion" in either case.

    Deliberately excludes fit_mode: it encoded the old strict/flexible
    ICP-fit *band* from the removed gate/D1-D7 scorer, and the current
    product scores purely on evidence (see evidence_scorer.py). An ICP is a
    seed for discovery, never a term in the Lead Score.
    """

    name: str | None = None
    industries: list[str] | None = None
    employee_min: int | None = None
    employee_max: int | None = None
    revenue_min_usd: int | None = None
    revenue_max_usd: int | None = None
    countries: list[str] | None = None
    technologies: list[str] | None = None
    buying_committee_personas: list[str] | None = None
    departments: list[str] | None = None

    @model_validator(mode="after")
    def _check_ranges(self):
        # A reversed range silently matches nothing, which reads as "no
        # companies like this exist" rather than "this ICP is impossible" -
        # reject it at the edge instead.
        if self.employee_min is not None and self.employee_max is not None:
            if self.employee_min > self.employee_max:
                raise ValueError("employee_min cannot be greater than employee_max")
        if self.revenue_min_usd is not None and self.revenue_max_usd is not None:
            if self.revenue_min_usd > self.revenue_max_usd:
                raise ValueError("revenue_min_usd cannot be greater than revenue_max_usd")
        return self


class IcpCreate(IcpBase):
    pass


class IcpOut(IcpBase):
    model_config = ConfigDict(from_attributes=True)

    icp_id: UUID
    workspace_id: UUID
    created_at: datetime | None = None
    updated_at: datetime | None = None


class IcpOptionsOut(BaseModel):
    """Vocabulary for the ICP form's pickers, served from the backend so the
    frontend never carries a second copy that can drift from what the data
    actually contains.

    This matters more than it looks: an earlier ICP form hardcoded invented
    industry labels ("Software & SaaS") that matched no real company, and a
    persona list missing 10 of the real values, so criteria silently selected
    nothing. Industries come from industry_sectors (the same mapping the
    Dashboard/Enterprise List segment on) and personas from the
    decision_maker.persona CHECK constraint.
    """

    industries: list[str]
    sectors: dict[str, list[str]]
    personas: list[str]
    departments: list[str]


# Mirrors lead_generation.MAX_TARGET. Duplicated as a literal rather than
# imported because schemas must not depend on services (services import
# schemas); tests/test_lead_generation.py asserts the two stay equal.
MAX_GENERATION_TARGET = 100


class GenerateLeadsIn(BaseModel):
    """Request body for POST /workspaces/{id}/icp/{icp_id}/generate.

    `target` is capped because generation is the one path that can create far
    more companies than anyone would upload by hand, and every company costs a
    verification search plus a full research pass. Precision beats volume here:
    a short verified list is worth more than a long padded one.
    """

    target: int = Field(default=25, ge=1, le=MAX_GENERATION_TARGET)

File: frontend/src/api/icp.ts
/* Three separate things share this module for historical reasons (the name
 * predates the split, and the shared types have many importers):
 *
 * - ICP CRUD (IcpOut/IcpCreate/listIcps/...) mirrors backend/app/routes/icp.py,
 *   backing the ICP page at /icp. An ICP here is a *seed for discovery* - it
 *   describes which companies to go find. It is NOT a filter on scoring and
 *   has no term in the Lead Score; see ICP_LEAD_GENERATION_INTENT.md.
 * - CompanyOut/DecisionMakerOut are general-purpose types several pages import.
 * - listImportBatches/ImportBatchOut and the job* helpers back Onboarding/
 *   Settings' upload history, which has no ICP dependency and hits
 *   backend/app/routes/icp_imports.py. */
import { apiDelete, apiGet, apiPost, apiPut } from "./client";

export type DecisionMakerOut = {
  decision_maker_id: string;
  company_id: string;
  zi_person_id: number;
  first_name: string | null;
  last_name: string | null;
  picture_url: string | null;
  job_title: string | null;
  department: string | null;
  years_of_experience: string | null;
  persona: string | null;
  email: string | null;
  phone: string | null;
  mobile_phone: string | null;
  linkedin_url: string | null;
};

export type CompanyOut = {
  company_id: string;
  zi_company_id: number;
  company_name: string;
  company_domain: string | null;
  company_type: string | null;
  company_status: string | null;
  is_verified: boolean | null;
  employee_count: number | null;
  employee_range: string | null;
  revenue_usd: number | null;
  revenue_range: string | null;
  ownership_type: string | null;
  founded_year: string | null;
  description: string | null;
  logo_url: string | null;
  city: string | null;
  state: string | null;
  country: string | null;
  continent: string | null;
  primary_industry: string[] | null;
  industries: string[] | null;
  linkedin_url: string | null;
  twitter_url: string | null;
  facebook_url: string | null;
  total_funding_amount: number | null;
  recent_funding_amount: number | null;
  recent_funding_date: string | null;
  company_funding: unknown;
  employee_growth: unknown;
  competitors: unknown;
  technologies: string[] | null;
  products: string[] | null;
  decision_makers?: DecisionMakerOut[];
};

export type ImportBatchOut = {
  import_batch_id: string;
  workspace_id: string | null;
  /* How this batch's companies came to exist. "generated" batches were
   * discovered from an ICP (icp_id names it) and arrive with no contacts, so
   * their Contact Access is 0 and their Lead Score is capped lower than an
   * uploaded company's - which is why the two are labelled rather than mixed
   * silently. */
  source: "upload" | "generated";
  icp_id: string | null; // set for generated batches, null for uploads
  icp_name: string | null;
  file_names: string[] | null;
  files_processed: number;
  total_rows: number;
  companies_ingested: number;
  signals_extracted: number;
  // New pipeline sales-status counts.
  sales_ready_count: number;
  high_priority_count: number;
  warm_count: number;
  monitor_count: number;
  low_priority_count: number;
  // Legacy read-only counters.
  matched_icp_count: number;
  active_count: number;
  nurture_count: number;
  // "pending" while research + scoring is still running in the background -
  // counts are 0 until this flips to "complete".
  scoring_status: "pending" | "complete";
  // Operational status of the background task (brief items 21, 23).
  research_status: "pending" | "complete" | "complete_with_warnings" | "failed";
  companies_researched: number;
  research_failure_count: number;
  llm_failure_count: number;
  scoring_failure_count: number;
  processing_started_at: string | null;
  processing_completed_at: string | null;
  processing_error: string | null;
  processing_warnings: string[] | null;
  created_at: string | null;
};

/* Every prospect upload in this workspace, newest first - the persisted audit
 * trail for the Settings prospect-data page and Enterprise List's per-upload
 * filter. Workspace-scoped, no ICP (brief section 7). */
export function listImportBatches(workspaceId: string): Promise<ImportBatchOut[]> {
  return apiGet<ImportBatchOut[]>(`/workspaces/${workspaceId}/imports`);
}

export type DeleteImportOut = {
  import_batch_id: string;
  file_names: string[];
  companies_deleted: number;
  /* Companies that ALSO belong to another upload and were therefore spared.
   * company_import_batch is a permanent many-to-many, so deleting an older
   * upload must not destroy companies a newer one still contains - surfacing
   * this lets the UI say what actually happened instead of implying all of
   * the batch's companies were removed. */
  companies_kept: number;
  buying_events_deleted: number;
};

/* Destructive and irreversible: deletes the upload AND the companies it
 * introduced, along with their buying events, scores and contacts. */
export function deleteImportBatch(
  workspaceId: string,
  importBatchId: string,
): Promise<DeleteImportOut> {
  return apiDelete<DeleteImportOut>(`/workspaces/${workspaceId}/imports/${importBatchId}`);
}

/* Per-company job monitoring for one upload (mirrors backend/app/schemas/job.py) -
 * "job" here is exactly this same ImportBatchOut/import_batch_id, just a
 * live per-company-status read view on top of it. Polled from the frontend
 * instead of keeping the upload request open. */

export type JobStatus = "queued" | "processing" | "partially_completed" | "completed" | "failed" | "cancelled";

export type CompanyJobStatus = "queued" | "researching" | "scoring" | "completed" | "retrying" | "failed" | "needs_review";

export type JobStatusOut = {
  job_id: string;
  status: JobStatus;
  total: number;
  queued: number;
  processing: number;
  completed: number;
  failed: number;
  needs_review: number;
  progress_percentage: number;
};

export type JobItemOut = {
  company_id: string;
  company_name: string;
  status: CompanyJobStatus;
  error_message: string | null;
  retry_count: number;
  started_at: string | null;
  completed_at: string | null;
};

export type JobItemsOut = {
  items: JobItemOut[];
  total: number;
  page: number;
  page_size: number;
};

export type RetryFailedOut = {
  retried_count: number;
  status: JobStatusOut;
};

export function getJobStatus(workspaceId: string, importBatchId: string): Promise<JobStatusOut> {
  return apiGet<JobStatusOut>(`/workspaces/${workspaceId}/imports/${importBatchId}`);
}

export function getJobItems(
  workspaceId: string,
  importBatchId: string,
  params: { page?: number; page_size?: number; status?: CompanyJobStatus } = {},
): Promise<JobItemsOut> {
  const query = new URLSearchParams();
  if (params.page) query.set("page", String(params.page));
  if (params.page_size) query.set("page_size", String(params.page_size));
  if (params.status) query.set("status", params.status);
  const qs = query.toString();
  return apiGet<JobItemsOut>(`/workspaces/${workspaceId}/imports/${importBatchId}/items${qs ? `?${qs}` : ""}`);
}

export function retryFailedJobItems(workspaceId: string, importBatchId: string): Promise<RetryFailedOut> {
  return apiPost<RetryFailedOut>(`/workspaces/${workspaceId}/imports/${importBatchId}/retry-failed`);
}

export function cancelJob(workspaceId: string, importBatchId: string): Promise<JobStatusOut> {
  return apiPost<JobStatusOut>(`/workspaces/${workspaceId}/imports/${importBatchId}/cancel`);
}

/* ── ICP definitions (mirrors backend/app/routes/icp.py) ──────────────────
 *
 * Every criterion is optional; an unset one means "no constraint". There is
 * deliberately no `fit_mode` and no "companies matching this ICP" call - both
 * belonged to the removed ICP-as-scoring-filter design. */

export type IcpOut = {
  icp_id: string;
  workspace_id: string;
  name: string | null;
  industries: string[] | null;
  employee_min: number | null;
  employee_max: number | null;
  revenue_min_usd: number | null;
  revenue_max_usd: number | null;
  countries: string[] | null;
  technologies: string[] | null;
  buying_committee_personas: string[] | null;
  departments: string[] | null;
  created_at: string | null;
  updated_at: string | null;
};

export type IcpCreate = {
  name?: string | null;
  industries?: string[] | null;
  employee_min?: number | null;
  employee_max?: number | null;
  revenue_min_usd?: number | null;
  revenue_max_usd?: number | null;
  countries?: string[] | null;
  technologies?: string[] | null;
  buying_committee_personas?: string[] | null;
  departments?: string[] | null;
};

/* Picker vocabulary, served by the backend so the form can't drift from the
 * values real data actually uses. `departments` is read from this org's own
 * contacts and is empty before any upload. */
export type IcpOptionsOut = {
  industries: string[];
  sectors: Record<string, string[]>;
  personas: string[];
  departments: string[];
};

export function listIcps(workspaceId: string): Promise<IcpOut[]> {
  return apiGet<IcpOut[]>(`/workspaces/${workspaceId}/icp`);
}

export function getIcpOptions(workspaceId: string): Promise<IcpOptionsOut> {
  return apiGet<IcpOptionsOut>(`/workspaces/${workspaceId}/icp/options`);
}

export function createIcp(workspaceId: string, payload: IcpCreate): Promise<IcpOut> {
  return apiPost<IcpOut>(`/workspaces/${workspaceId}/icp`, payload);
}

/* Full-replace (PUT): the edit form submits every field, so an omitted field
 * means "clear this criterion", not "leave it unchanged". */
export function updateIcp(workspaceId: string, icpId: string, payload: IcpCreate): Promise<IcpOut> {
  return apiPut<IcpOut>(`/workspaces/${workspaceId}/icp/${icpId}`, payload);
}

/* Upload history survives: icp_import_batch.icp_id is ON DELETE SET NULL, so
 * past uploads keep their counts and only lose the ICP link. Companies,
 * buying events and scores are organisation-scoped and untouched. */
export function deleteIcp(workspaceId: string, icpId: string): Promise<void> {
  return apiDelete<void>(`/workspaces/${workspaceId}/icp/${icpId}`);
}

/* Discovers new companies from an ICP: an LLM proposes candidates, each is
 * verified against live web search, and only the verified ones become
 * companies. Returns the SAME ImportBatchOut an upload does, with
 * scoring_status "pending" - poll the existing job endpoints (getJobStatus /
 * getJobItems / retryFailedJobItems / cancelJob) exactly as for an upload.
 *
 * Slow by nature (one search per candidate) and can legitimately fail with:
 *   503 - the LLM or search service isn't configured
 *   422 - nothing new could be verified (all invented, or all already owned) */
export function generateLeads(
  workspaceId: string,
  icpId: string,
  target: number,
): Promise<ImportBatchOut> {
  return apiPost<ImportBatchOut>(`/workspaces/${workspaceId}/icp/${icpId}/generate`, { target });
}
